File: SumRoottoLeafNumbers.java
// http://oj.leetcode.com/problems/sum-root-to-leaf-numbers/

/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

import java.util.*;
 
public class Solution {
	private int res;
	public int sumNumbers(TreeNode root) {
		res = 0;
		sumNumbersHelper(root, 0);
		return res;
	}
	
	private void sumNumbersHelper(TreeNode n, int num) {
		if (n == null) {
			return;
		}
		num += n.val;
		if (n.left == null && n.right == null) {
			res += num;
		}
		sumNumbersHelper(n.left, num * 10);
		sumNumbersHelper(n.right, num * 10);
	}
}

File: sum_root_to_leaf_numbers.js
// https://leetcode.com/problems/sum-root-to-leaf-numbers/

/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var sumNumbers = function(root) {
    var recur = function(root, sum) {
        // root cannot be null

        var newVal = sum * 10 + root.val;
        if (!root.left && !root.right) {
            // leaf
            return newVal;
        }
        return (root.left ? recur(root.left, newVal) : 0)
             + (root.right ? recur(root.right, newVal) : 0);
    };

    if (root === null) { return 0; }
    return recur(root, 0);
};
